core/src/main/java/smile/classification/IsotonicRegressionScaling.java

```diff
@@ -0,0 +1,178 @@
+        initStepwiseConstFuncs(scores, y);
+
+        StepwiseConsFunc current = startConstantFuncs.next;
+
+        while (current != null) {
+            StepwiseConsFunc previous = current.previous;
+            while (previous != null) {
+                if (previous.getVal() >= current.getVal()) {
+                    int weight = previous.getWeight() + current.getWeight();
+                    double val = (previous.getWeight() * previous.getVal() + current.getWeight() * current.getVal())
+                            / weight;
+                    StepwiseConsFunc newFunc = new StepwiseConsFunc(previous.lowScore, current.highScore, val);
+                    newFunc.setWeight(weight);
+
+                    newFunc.next = current.next;
+                    if (newFunc.next != null) {
+                        newFunc.next.previous = newFunc;
+                    }
+
+                    newFunc.previous = previous.previous;
+                    if (newFunc.previous != null) {
+                        newFunc.previous.next = newFunc;
+                    }
+
+                    current = newFunc;
+                    previous = current.previous;
+                } else {
+                    previous = previous.previous;
+                }
+            }
+            current = current.next;
+        }
+    }
+
+    private void initStepwiseConstFuncs(double[] scores, int[] y) {
+        double[] sortedScores = Arrays.copyOf(scores, scores.length);
+        int[] sortedY = Arrays.copyOf(y, y.length);
+
+        QuickSort.sort(sortedScores, sortedY, sortedScores.length);
+
+        StepwiseConsFunc previous = new StepwiseConsFunc(sortedScores[0], sortedScores[0], (double) sortedY[0]);
+        startConstantFuncs = previous;
+
+        for (int i = 1; i < sortedScores.length; i++) {
+            StepwiseConsFunc init = new StepwiseConsFunc(sortedScores[i], sortedScores[i], (double) sortedY[i]);
+            init.setPrevious(previous);
+            previous.setNext(init);
+            previous = init;
+        }
+    }
+
+    /**
+     * Returns the posterior probability estimate P(y = 1 | x).
+     *
+     * @param y
+     *            the binary classifier output score.
+     * @return the estimated probability.
+     */
+    public double predict(double y) {
+        double ret = Double.NaN;
+        StepwiseConsFunc func = startConstantFuncs;
+        while (func != null) {
+            if (func.matchValue(y)) {
+                ret = func.getVal();
+                break;
+            }
+            func = func.next;
+        }
+        if (ret == Double.NaN) {
+            throw new IllegalArgumentException("fail to get the posteriori probability for given prediction: " + y);
+        }
+        return ret;
+    }
+
+    /**
+     * Estimates the multiclass probabilies.
+     */
+    public static void multiclass(int k, double[][] r, double[] p) {
+        PlattScaling.multiclass(k, r, p);
```

Comment:
Can we do this? Do you have a reference?

Reply:
No. Reference for platt does not mention multi-class calculation neither: https://www.researchgate.net/profile/John_Platt/publication/2594015_Probabilistic_Outputs_for_Support_Vector_Machines_and_Comparisons_to_Regularized_Likelihood_Methods/links/004635154cff5262d6000000.pdf